SQC-652 Implement Hyperdrive TLS proxy in Miniflare

Ltadrian opened the pull request:
Fixes SQC-652.

This fixes the problem of when a user runs 
`npx wrangler dev` 
with a Hyperdrive binding and `localConnectionString` pointing to a database that requires TLS. In such cases, the command will result in the following TLS error:
```
{ "error": "connection is insecure (try using `sslmode=require`)" }
```

This change adds a proxy layer to Miniflare and introduces a proxy service for Hyperdrive bindings. The proxy layer attempts to establish a TLS connection with the appropriate database based on the specified sslmode parameter.

This allows users to simply set `sslmode=require` in their `localConnectionString`, enabling their Hyperdrive binding to automatically upgrade to TLS instead of requiring them to manually configure TLS settings in their database driver. (Note: this still will not work without additional changes to the binding service.)

---

<!--
Please don't delete the checkboxes <3
The following selections do not need to be completed if this PR only contains changes to .md files
-->

- Tests
  - [X] Tests included
  - [ ] Tests not necessary because:
- Public documentation
  - [X] Cloudflare docs PR(s): https://github.com/cloudflare/cloudflare-docs/pull/26545
  - [ ] Documentation not necessary because:
- Wrangler V3 Backport
  - [ ] Wrangler PR: <!--e.g. <https://github.com/cloudflare/workers-sdk/pull/>...-->
  - [X] Not necessary because: New feature behavior

<!--
Have you read our [Contributing guide](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md)?
In particular, for non-trivial changes, please always engage on the issue or create a discussion or feature request issue first before writing your code.
-->

Ltadrian commented on packages/miniflare/src/plugins/hyperdrive/index.ts:
```diff
@@ -115,15 +115,56 @@ export const HYPERDRIVE_PLUGIN: Plugin<typeof HyperdriveInputOptionsSchema> = {
 				name: `${HYPERDRIVE_PLUGIN_NAME}:${name}`,
 				external: {
-					address: `${url.hostname}:${getPort(url)}`,
+					address: `localhost:${proxyPort}`,
```
I've been testing this locally and so far for me using `localhost` works whereas setting `127.0.0.1` seems to fail to connect to the integration test server. Not sure exactly why but maybe there is some IPv4/IPv6 conflict?

petebacondarwin replied:
Cool. No probs. Thanks for confirming

Ltadrian commented on packages/wrangler/e2e/get-platform-proxy.test.ts:
```diff
@@ -365,70 +367,210 @@ describe("getPlatformProxy()", () => {
-						compatibility_date = "2024-08-20"
-						compatibility_flags = ["nodejs_compat"]
+		it.skipIf(
+			// in CI this test fails for windows because of ECONNRESET issues
```
Testing this on my windows machine and I am able to reproduce the ECONNRESET errors, 
```
Error: read ECONNRESET

at TCP.onStreamRead (node:internal/stream_base_commons:216:20)
Emitted 'error' event on Socket instance at:
  at emitErrorNT (node:internal/streams/destroy:170:8)
  at emitErrorCloseNT (node:internal/streams/destroy:129:3)
  at process.processTicksAndRejections (node:internal/process/task_queues:89:21) {
    errno: -4077,
    code: 'ECONNRESET',
    syscall: 'read'
}
```

Ltadrian commented:
@petebacondarwin think I could get a re-review on this? I fixed some more tests and cleaned up some issues when testing more edge cases. I think currently the tests failing are flaky tests?

vicb commented on packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts:
```diff
@@ -0,0 +1,349 @@
+	// Skip connection_id, auth_plugin_data_part_1, filler
+	offset += 4;
+	offset += 8;
+	offset += 1;
```
Use constants with comments instead of magic numbers

ReppCodes commented on packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts:
```diff
@@ -0,0 +1,419 @@
+		const sslmodePrefer = sslmode === "prefer";
+		if (sslmodePrefer || sslmodeRequire) {
+			try {
+				if (scheme === "postgres" || scheme === "postgresql") {
```
Might be nice to have the supported schemes as a const array of magic up top. I doubt we'll add more, but for the sake of maintainability might be good to do.

ReppCodes commented on packages/wrangler/e2e/dev.test.ts:
```diff
@@ -588,9 +588,23 @@ describe.each([{ cmd: "wrangler dev" }])(
 
 describe("hyperdrive dev tests", () => {
 	let server: nodeNet.Server;
+	const sslRequestPacket = Buffer.from([
```
Ditto magic constant comment here.

Ltadrian commented on packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts:
```diff
@@ -0,0 +1,419 @@
+	sslmodeRequire: boolean
+) {
+	// Send Postgres sslrequest bytes
+	const sslRequestPacket = Buffer.from([
```
Tried creating a shared constant but it's a bit hard to do that in this repo with disjoint modules. I've created the shared const in a few places, miniflare/wrangle e2e/vitest-pool-workers-example modules

vicb commented on packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts:
```diff
@@ -0,0 +1,427 @@
+
+	/** Disposes of the proxy servers when shutting down the worker.*/
+	async dispose(): Promise<void> {
+		await Promise.all(
```
`all` will reject as soon as one of the Promise rejects.
`allSettled` will wait for all Promises to settle (either successfully or not).

```suggestion
		await Promise.allSettled(
```

Would there be any value in adding a debug log on error?

vicb commented on packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts:
```diff
@@ -0,0 +1,427 @@
+	return;
+}
+
+/** Helper function to create and wait for plain TCP connection */
```
```suggestion
/** Create and wait for plain TCP connection */
```

vicb commented on packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts:
```diff
@@ -0,0 +1,427 @@
+	return dbSocket;
+}
+
+/** Helper function to set up TLS connection with pipes and error handlers */
```
```suggestion
/** Set up TLS connection with pipes and error handlers */
```

vicb commented on packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts:
```diff
@@ -0,0 +1,427 @@
+}
+
+/** Helper functions for net library */
+function writeAsync(
```
suggestion, it looks like this function is often called with `Buffer.from(array)`, would it makes sense that bytes could also be an array and turned into a Buffer here if needed?

vicb commented on packages/miniflare/src/plugins/hyperdrive/index.ts:
```diff
@@ -115,15 +115,66 @@ export const HYPERDRIVE_PLUGIN: Plugin<typeof HyperdriveInputOptionsSchema> = {
+
+// Postgres sslmode docs: https://www.postgresql.org/docs/current/libpq-ssl.html
+// MySQL ssl-mode docs: https://dev.mysql.com/doc/refman/8.4/en/using-encrypted-connections.html
+function parseSslMode(url: URL, scheme: string): string {
```
Question: should search params always be normalized (to either casing)?

It's only done for some code paths.

vicb commented on packages/wrangler/src/__tests__/api/startDevWorker/LocalRuntimeController.test.ts:
```diff
@@ -1047,9 +1047,21 @@ describe("LocalRuntimeController", () => {
+						// echo back result
+						socket.write(chunk);
+					}
+				});
```
I feel like I have been seeing this code many times at this point in the review.

Is there an opportunity to factor some code ?

vicb reviewed (approved):
LGTM.

I have added mostly minor comments / suggestions.

If one thing, I feel like there is quite some code duplication across the PR.
